frontend/apps/reader/modules/readerhighlight.lua

```diff
@@ -1773,18 +1772,53 @@ function ReaderHighlight:saveHighlight(extend_to_sentence)
             bookmark_item.chapter = chapter_name
             self.ui.bookmark:addBookmark(bookmark_item)
         end
-        if self.selected_text.pboxes then
-            self:exportToDocument(page, hl_item)
-        end
+        self:writePdfAnnotation("save", page, hl_item)
         return page, #self.view.highlight.saved[page]
     end
 end
 
-function ReaderHighlight:exportToDocument(page, item)
-    local setting = G_reader_settings:readSetting("save_document")
-    if setting == "disable" then return end
-    logger.dbg("export highlight to document", item)
-    local can_write = self.ui.document:saveHighlight(page, item)
+function ReaderHighlight:writePdfAnnotation(action, page, hl_or_bm, content)
+    if self.ui.rolling or G_reader_settings:readSetting("save_document") == "disable" then
+        return
+    end
+    logger.dbg("write to pdf document", action, hl_or_bm)
+    local item
+    if hl_or_bm.highlighted then -- called from bookmarks to write bookmark note
+        item = self:getBookmarkPboxes(hl_or_bm)
+    else
+        item = hl_or_bm
+    end
+    local function doAction(_action, _page, _item, _content)
+        if _action == "save" then
+            return self.ui.document:saveHighlight(_page, _item)
+        elseif _action == "delete" then
+            return self.ui.document:deleteHighlight(_page, _item)
+        elseif _action == "content" then
+            return self.ui.document:updateHighlightContents(_page, _item, _content)
+        end
+    end
+    local can_write
+    if item.pos0.page == item.pos1.page then
+        can_write = doAction(action, page, item, content)
+    else -- multi-page highlight
+        local is_reflow = self.ui.document.configurable.text_wrap
+        for hl_page = item.pos0.page, item.pos1.page do
+            self.ui.document.configurable.text_wrap = 0
+            local hl_part = self:getSavedExtendedHighlightPage(item, hl_page)
+            self.ui.document.configurable.text_wrap = is_reflow
+            can_write = doAction(action, hl_page, hl_part, content)
+            if can_write == false
+                    -- content is written to the first page annotation
+                    or (action == "content" and hl_page == item.pos0.page) then
```

Comment:
(Might be just me, but not fond of the comment in a line in the middle of conditions. I would put it above, or at some end of line.)